### PR TITLE
Introduce tee subsystem support and OP-TEE OS driver

### PR DESCRIPTION
**Pull request by oleksiimoisieiev**
This PR introduces Tee subsystem interface and OP-TEE driver implementation for Zephyr.
The issue for integrating external source code: https://github.com/zephyrproject-rtos/zephyr/issues/61031

TEE subsystem implements generic interface to the Trusted Execution Environment.
OP-TEE driver implementation provides an interface to the OP-TEE TrustZone from the Normal space
to start sessions and request functions of the TA.
    
- targets ARM and ARM64;
- use SMC to connect to OP-TEE;
- accepts requests on privileged and unprivileged device.
    
See [docs](https://optee.readthedocs.io/en/latest/) for additional information.

Unit tests for optee driver are also introduced in terms of this PR. These tests should be built for native_posix_64 architecture and doesn't require connection to the real OP-TEE OS because of using mocked arm_smccc_smc and arm_smccc_hvc calls.

Testing with real OP-TEE OS can be performed by using ported [zephyr_optee_client] library and [zephyr_optee_test] application.

[zephyr_optee_client] is ported [optee_client] library from v3.18.0. It implements:
- tee_supplicant
- libckteec
- libteec
- libseteec

[zephyr_optee_test] is ported [optee_test] from v3.18.0. It was adopted to use Zephyr Ztest subsystem so successful run should look like [this](https://pastebin.com/fwPhMCT3).
Please consult with README.md to get build details. It requires prebuilt TA bins which can be taken from optee_os and optee_test build.

As was mentioned, [zephyr_optee_client] and [zephyr_optee_test] were ported from v3.18.0, so the target board should also have [optee_os] v3.18.0.

[zephyr_optee_test]https://github.com/oleksiimoisieiev/zephyr-optee-test/tree/tests_main
[zephyr_optee_client]https://github.com/oleksiimoisieiev/zephyr-optee-client/tree/client_main
[optee_client] https://github.com/OP-TEE/optee_client
[optee_test] https://github.com/OP-TEE/optee_test
[optee_os] https://github.com/OP-TEE/optee_os

### CODE REVIEW

**Comment by lorc** [comment]
Hi @jenswi-linaro, @jforissier,

I believe, you will be interested in this PR.

**Comment by oleksiimoisieiev** [comment]
As I can see Scancode was failed with the following error message:
>  Error: * scan/include/zephyr/drivers/tee.h has invalid license: bsd-simplified

It complains that the header includes Linaro copyright notice. This is open header, which is supplied as part of the [optee_client](https://github.com/OP-TEE/optee_client/blob/master/libteec/include/linux/tee.h), And I can't just drop this notice because it's a direct requirement.
As far as I know bsd-simplified is compatible with Apache-2.0 which is used by Zephyr. Could you please advise what should be done in this case?

**Comment by carlescufi** [comment]
@oleksiimoisieiev thanks for the proposal! I can't figure out easily if you are pulling in external source code or not in this Pull Request. If you are then you will need to follow the process outlined below:

https://docs.zephyrproject.org/latest/contribute/external.html#

**Comment by carlescufi** [comment]
> It complains that the header includes Linaro copyright notice. This is open header, which is supplied as part of the [optee_client](https://github.com/OP-TEE/optee_client/blob/master/libteec/include/linux/tee.h), And I can't just drop this notice because it's a direct requirement.
> As far as I know bsd-simplified is compatible with Apache-2.0 which is used by Zephyr. Could you please advise what should be done in this case?

Sorry, missed this comment before I posted mine. This means you are pulling an external header, so you should follow the process I linked above.

**Comment by carlescufi** [comment]
@microbuilder is there someone at Linaro who could help review this?

**Comment by carlescufi** [comment]
Also, @oleksiimoisieiev can you clarify what the difference is between OP-TEE and TF-A and why we need both in Zephyr (we already have TF-A as a module).

**Comment by oleksiimoisieiev** [comment]
> Also, @oleksiimoisieiev can you clarify what the difference is between OP-TEE and TF-A and why we need both in Zephyr (we already have TF-A as a module).

Hi @carlescufi,
[TF-A](https://trustedfirmware-a.readthedocs.io/en/latest/about/features.html) implement mostly boot flow and system resource management in Secure World and provides interfaces to the Normal world to control those resources, such as SCMI. And also responsible for loading Stage-2 bootladers, such as u-boot and OPTEE. 
For example [atf](https://github.com/ARM-software/arm-trusted-firmware) acts like Stage-1 loader.
[OPTEE-OS](https://optee.readthedocs.io/en/latest/general/index.html) provides trusted execution environment(TEE) for Trusted Applications which could be run in Secure World. It's goal is to provide isolation from the non-secure OS and protects loaded  TA's from each other.

**Comment by oleksiimoisieiev** [comment]
Created TSC request for the external sources: https://github.com/zephyrproject-rtos/zephyr/issues/61031

**Review by galak — COMMENTED** [review]
Is linaro,optee-tz devicetree defined in LInux?

How does this related to drivers/sip_svc/?

**Comment by oleksiimoisieiev** [comment]
> Is linaro,optee-tz devicetree defined in LInux?


It is. See [binfings](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/arm/firmware/linaro%2Coptee-tz.yaml)


> How does this related to drivers/sip_svc/?

Could you please clarify what is drivers/sip_svc? Couldn't find it in Linux kernel source code.

**Review by carlocaione — DISMISSED** [review]
In general looks fine to me but when we introduced PSCI we also added a small SMCCC library `arch/arm64/core/smccc-call.S` and `./include/zephyr/arch/arm64/arm-smccc.h` and I'd like to see also OPTEE use that one.

File: tests/drivers/tee/optee/src/main.c
Comment (oleksiimoisieiev) [reply tests/drivers/tee/optee/src/main.c:67]: This was done for the unit testing. optee unit tests are built on posix_native_64 platform, which doesn't have arm_smccc so I could replace arm_smccc_smc call with my own implementation to check the parameters and avoid calling optee_os. For integration tests with real optee_os zephyr_optee_test application was created as was mentioned in the description.

File: drivers/tee/optee/Kconfig
Comment (oleksiimoisieiev) [reply drivers/tee/optee/Kconfig]: fixed

**Comment by povergoing** [comment]
Just out of curiosity, do you plan to support FF-A?

**Comment by oleksiimoisieiev** [comment]
> Just out of curiosity, do you plan to support FF-A?

We haven't imlemented FFA interface yet.

File: drivers/tee/optee/optee_msg.h
Comment (carlocaione) [inline drivers/tee/optee/optee_msg.h:12]: What's preventing to do that in this PR?

**Review by carlocaione — DISMISSED** [review]
LGTM but CI is failing on the license. @nashif @carlescufi any idea about that?

**Comment by xakep-amatop** [comment]
> > How does this related to drivers/sip_svc/?
> 
> Could you please clarify what is drivers/sip_svc? Couldn't find it in Linux kernel source code.

I believe @galak was referring to the Zephyr `sip_svc` driver, which can be found here: https://elixir.bootlin.com/zephyr/latest/source/drivers/sip_svc

**Comment by oleksiimoisieiev** [comment]
Sorry for the delayed answer.
I've rebased PR and fixed conflicts. Is there any progress with TSC?

> I believe @galak was referring to the Zephyr `sip_svc` driver, which can be found here: https://elixir.bootlin.com/zephyr/latest/source/drivers/sip_svc

Looks like sip_svc_driver is not what optee needs.

**Comment by oleksiimoisieiev** [comment]
Hello @carlescufi.
Is there any updates about this pull request? I can see TSC tag is already removed.

**Comment by nashif** [comment]
DNM because we need approval for the incoming license from the board.

**Comment by d3zd3z** [comment]
Any updates on the status of this, e.g., the license review?

**Comment by oleksiimoisieiev** [comment]
Hi @nashif @carlescufi. As I can see TSC reqeuest https://github.com/zephyrproject-rtos/zephyr/issues/61031 was moved to the Done state. Do you have any plans about merging this PR?

**Comment by nashif** [comment]
> Hi @nashif @carlescufi. As I can see TSC reqeuest #61031 was moved to the Done state. Do you have any plans about merging this PR?

license request was approved by the board.

**Comment by oleksiimoisieiev** [comment]
Hi @nashif @carlescufi. Do you have any updates regarding this PR?

**Review by ithinuel — COMMENTED** [review]
Since from the PR description, it seems that this is tied to a specific version of op-tee, wouldn’t it make sense to have an error log (or at least a warning) if there’s a version mismatch, eg when calling `tee_open_session` ?

@carlescufi

**Comment by carlescufi** [comment]
> Since from the PR description, it seems that this is tied to a specific version of op-tee, wouldn’t it make sense to have an error log (or at least a warning) if there’s a version mismatch, eg when calling `tee_open_session` ?
> 
> @carlescufi

That sounds reasonable to me, but it's a question for @oleksiimoisieiev

**Comment by oleksiimoisieiev** [comment]
> Since from the PR description, it seems that this is tied to a specific version of op-tee, wouldn’t it make sense to have an error log (or at least a warning) if there’s a version mismatch, eg when calling `tee_open_session` ?
> 
> @carlescufi

You are right. 
There is a room for improvement and we continue the development.
There are some PRs (alrealy merged in our repos) that does improvements, such as request OS revision and caps:
https://github.com/xen-troops/zephyr/pull/65
https://github.com/xen-troops/zephyr/pull/70
or shm caching and tmpmem handling
https://github.com/xen-troops/zephyr/pull/71
https://github.com/xen-troops/zephyr/pull/66

I wanted to add this functionality as a separate PR because this one already passed a couple of review iterations.
The only thing I'd like to add to this implementation is the following fix:
https://github.com/xen-troops/zephyr/pull/83
This fixing address conversion by using
```
arch_page_phys_get
```
function. We have met this issue when using pkcs11 library which allocates buffer in dynamic memory.
I haven't introduced it yet because current version works with optee-tests and pkcs11 library can be modified to use static buffer instead of dynamic.

So as a summary: my idea is to introduce functionality step-by-step and adding new features as a separate PRs to make them easier to review. 
What do you think about this?

**Comment by carlescufi** [comment]
@ithinuel please take a look at [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/60680#issuecomment-2165085723)

**Review by ithinuel — DISMISSED** [review]

**Review by carlescufi — DISMISSED** [review]

**Comment by oleksiimoisieiev** [comment]
Please do not merge this PR. I've just found a strange issue with corrupted stack on zephyr while running zephyr-optee-test. This probably connected to the memory allocation but I want to be sure. Also, I wanted to create a detailed description about how to build and start zephyr-optee-test on the real board and put it here in the comment section.

**Comment by kartben** [comment]
adding DNM as per @oleksiimoisieiev latest comment to avoid fly-by merge

**Comment by nashif** [comment]
@oleksiimoisieiev any updates?

File: drivers/tee/optee/optee.c
Comment (nashif) [inline drivers/tee/optee/optee.c:168]: z_mem_phys_addr was deprecated, please use k_mem_phys_addr instead.

**Review by nashif — CHANGES_REQUESTED** [review]

**Comment by oleksiimoisieiev** [comment]
> @oleksiimoisieiev any updates?

Hi @nashif. We have found a couple of problems while running zephyr_optee_test. One is the corner case with handling multithread access to optee. I'm fixing it right now and going to add to this PR. The second issue was with the stack size.
When big buffer is allocating in the test_case on stack:
```
ZTEST(...) {
   uint8_t buf [32 * 1024];
   ...
}
```
there is risk to corrupt global variables: test_result and failed_expectation during test execution.
test_result corruption doesn't produce any errors because it's overwritten in `test_cb` function right after `run_test_functions` call:
```
static void test_cb(void *a, void *b, void *c)
{
       .....
	run_test_functions(suite, test, c);
	__ztest_set_test_result(ZTEST_RESULT_PASS);
}
```
But failed_expectation corruption causes test to be failed even if it was passed without errors.

Currently we have fixed this by moving 
```
   uint8_t buf [32 * 1024];
```
to the global section, but I think there should be some mechanism like placing these variables in some `test section` to avoid such errors in future.

**Comment by ithinuel** [comment]
@oleksiimoisieiev you can alter the stack size configuration for the test threads with `CONFIG_ZTEST_STACK_SIZE`. A few tests place that in their `prj.conf` file.

File: drivers/tee/optee/optee.c
Comment (oleksiimoisieiev) [reply drivers/tee/optee/optee.c:168]: thank you. fixed.

**Comment by oleksiimoisieiev** [comment]
Hello @ithinuel  @nashif,

I have updated PR to v2.
Changes v1 -> v2

 - used k_mem_phys_addr instead of z_mem_phys_addr
 - fixed issue with normal_call responce on the unit test
 - handling exchange capabilities call to the unit tests
 - add GET_THREAD_COUNT call to the optee driver
    This call requests maximum threads supported by optee and limits
    amount of parallel calls to optee.
    
Steps to start zephyr-optee-test. Tested on Renesas R-car Gen3 board (r8a77951):

## Build optee-os:

url = git@github.com:OP-TEE/optee_os.git
rev = 3.18.0

Building:
```
make -j 17 \
 PLATFORM=rcar \
 PLATFORM_FLAVOR=salvator_m3_2x4g \
 CROSS_COMPILE=aarch64-linux-gnu- \
 CROSS_COMPILE64=aarch64-linux-gnu- \
 CFG_TEE_TA_LOG_LEVEL=4 \
 CFG_TEE_CORE_LOG_LEVEL=4 \
 CFG_ARM64_core=y \
 CFG_SYSTEM_PTA=y \
 CFG_RCAR_UART=103 \
 CFG_DEBUG=y \
 CFG_CORE_ASYNC_NOTIF=y \
 CFG_CORE_ASYNC_NOTIF_GIC_INTID=45 \
 CFG_ENABLE_EMBEDDED_TESTS=y $1
```

## Build optee-test:

url = git@github.com:OP-TEE/optee_test.git
rev = 3.18.0

Building:
```
make -j 8      TA_DEV_KIT_DIR=<devkit>/optee/export-user_ta_arm64     OPTEE_CLIENT_EXPORT=<target>/optee/usr     TEEC_EXPORT=<target>/optee/usr     CROSS_COMPILE_HOST=aarch64-linux-gnu-     CROSS_COMPILE_TA=aarch64-linux-gnu-     V=1  ta
```

## Building zephyr

url = git@github.com:oleksiimoisieiev/zephyr-optee-test.git
rev = zephyr_pr_migr

then execute command
```
west update
```

Switch zephyr-optee-client to the following tag:
```
v0.0.1
```

Switch zephyr to this pull request latest revision:
```
optee_driver_sq
```
from my fork.

Create folder `prebuilt` in zeprhyr-optee-test
Copy TA's from `export-user_ta_arm64` directories created after building optee-os and optee-test.

Build application:
```
west build -b  rcar_h3ulcb/r8a77951/a57 -p always -- -DTA_DEPLOY_DIR=$(pwd)/prebuilt
```

After flashing tee.srec created after building optee-os and starting the board with zephyr.bin tests will be started.

## Problems solved

The first problems were with global variables corruption:
 - test 6018 corrupted 
   It was allocating a big buffer as function local variable which corrupted
   failed_expectation and test_result globals.
   With test_result the corruction isn't significant because it is
   overwritten right after test finish. But failed_expectation corruption
   causing weird behaviour when Test was passed but it is marked as failed.
   Increasing CONFIG_MAIN_STACK_SIZE to 64000 and CONFIG_ZTEST_STACK_SIZE to 32000
   did not help so I've move buffer to the global section
 - test_1014 of pkcs11. g_decrypt local variable that should store decrypt attribute was
   always FALSE while optee-os was returning TRUE. Was fixed by moving variables to the local
   section.
 - test 6016 with multithreading access. Was fixed by handling GET_THREAD_COUNT call.
 - Added CONFIG_FS_LITTLEFS_NUM_FILES configuration because zephyr_optee_test was
   trying to create 5 files which doesn't fit the default 4 files max.

**Review by nashif — APPROVED** [review]

**Comment by nashif** [comment]
@ithinuel  can you please revisit?

**Review by carlescufi — APPROVED** [review]

**Review by ithinuel — APPROVED** [review]

**Pull request merged**
